src/masternodes/order.h

```diff
@@ -0,0 +1,222 @@
+#ifndef DEFI_MASTERNODES_ORDER_H
+#define DEFI_MASTERNODES_ORDER_H
+
+#include <amount.h>
+#include <uint256.h>
+
+#include <flushablestorage.h>
+#include <masternodes/res.h>
+
+class COrder
+{
+public:
+    static const int DEFAULT_ORDER_EXPIRY = 2880;
+
+    //! basic properties
+    std::string ownerAddress;
+    std::string tokenFrom;
+    std::string tokenTo;
+    DCT_ID idTokenFrom;
+    DCT_ID idTokenTo;
+    CAmount amountFrom;
+    CAmount orderPrice;
+    uint32_t expiry;
+
+    COrder()
+        : ownerAddress("")
+        , tokenFrom("")
+        , tokenTo("")
+        , idTokenFrom({0})
+        , idTokenTo({0})
+        , amountFrom(0)
+        , orderPrice(0)
+        , expiry(DEFAULT_ORDER_EXPIRY)
+    {}
+    virtual ~COrder() = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITE(ownerAddress);
+        READWRITE(tokenFrom);
+        READWRITE(tokenTo);
+        READWRITE(VARINT(idTokenFrom.v));
+        READWRITE(VARINT(idTokenTo.v));
+        READWRITE(amountFrom);
+        READWRITE(orderPrice);
+        READWRITE(expiry);
+    }
+};
+
+class COrderImplemetation : public COrder
+{
+public:
+    //! tx related properties
+    uint256 creationTx;
+    uint256 closeTx;
+    uint32_t creationHeight; 
+    uint32_t closeHeight;
+
+    COrderImplemetation()
+        : COrder()
+        , creationTx()
+        , closeTx()
+        , creationHeight(-1)
+        , closeHeight(-1)
+    {}
+    ~COrderImplemetation() override = default;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action) {
+        READWRITEAS(COrder, *this);
+        READWRITE(creationTx);
+        READWRITE(closeTx);
+        READWRITE(creationHeight);
+        READWRITE(closeHeight);
+    }
+};
+
+class CFulfillOrder
+{
+public:
+    //! basic properties
+    std::string ownerAddress;
+    uint256 orderTx;
```

Comment:
Should `CFulfillOrder` have two orders? one is sell order and one is buy order

Reply:
CFulfillOrder is a class for fulfill order tx, and it refers on orderTx which is a txid on order which which it fulfills. Basically you can think of it as order is a let say sell order and fulfill order is buy order or vice versa.